File: sdks/java-sdk/src/main/java/com/quantaureum/sdk/types/Hash.java
// Quantaureum Java SDK source, version 1.0.0.
package com.quantaureum.sdk.types;

import com.quantaureum.sdk.utils.Hex;

import java.util.Arrays;

/**
 * Represents a 32-byte hash value.
 *
 * <p>This class is immutable and thread-safe.
 *
 * <h2>Usage Examples</h2>
 * <pre>{@code
 * // Create from hex string
 * Hash hash = Hash.fromHex("0x1234...");
 *
 * // Create from bytes
 * Hash hash = Hash.fromBytes(bytes);
 *
 * // Get hex representation
 * String hex = hash.toHex();
 * }</pre>
 */
public final class Hash {

    /** The zero hash (0x0000...0000) */
    public static final Hash ZERO = new Hash(new byte[32]);

    /** Hash length in bytes */
    public static final int LENGTH = 32;

    private final byte[] bytes;

    private Hash(byte[] bytes) {
        this.bytes = bytes.clone();
    }

    /**
     * Creates a Hash from a byte array.
     *
     * @param bytes the 32-byte hash
     * @return a Hash instance
     * @throws IllegalArgumentException if bytes is null or not 32 bytes
     */
    public static Hash fromBytes(byte[] bytes) {
        if (bytes == null) {
            throw new IllegalArgumentException("Hash bytes cannot be null");
        }
        if (bytes.length != LENGTH) {
            throw new IllegalArgumentException(
                "Hash must be " + LENGTH + " bytes, got " + bytes.length);
        }
        return new Hash(bytes);
    }

    /**
     * Creates a Hash from a hex string.
     *
     * @param hex the hex-encoded hash (with or without 0x prefix)
     * @return a Hash instance
     * @throws IllegalArgumentException if hex is invalid
     */
    public static Hash fromHex(String hex) {
        if (hex == null || hex.isEmpty()) {
            throw new IllegalArgumentException("Hash hex cannot be null or empty");
        }

        String cleanHex = hex.startsWith("0x") || hex.startsWith("0X")
            ? hex.substring(2) : hex;

        if (cleanHex.length() != LENGTH * 2) {
            throw new IllegalArgumentException(
                "Hash hex must be " + (LENGTH * 2) + " characters, got " + cleanHex.length());
        }

        byte[] bytes = Hex.toBytes(cleanHex);
        return new Hash(bytes);
    }

    /**
     * Returns the hash as a byte array.
     *
     * @return a copy of the hash bytes
     */
    public byte[] toBytes() {
        return bytes.clone();
    }

    /**
     * Returns the hash as a lowercase hex string with 0x prefix.
     *
     * @return the hex-encoded hash
     */
    public String toHex() {
        return "0x" + Hex.toHexString(bytes);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Hash hash = (Hash) o;
        return Arrays.equals(bytes, hash.bytes);
    }

    @Override
    public int hashCode() {
        return Arrays.hashCode(bytes);
    }

    @Override
    public String toString() {
        return toHex();
    }
}

File: sdks/cpp-sdk/src/types/hash.cpp
// Quantaureum C++ SDK source, version 1.0.0.
#include "quantaureum/types/hash.hpp"
#include "quantaureum/utils/hex.hpp"
#include <stdexcept>
#include <algorithm>

namespace quantaureum {

Hash::Hash() : bytes_{} {}

Hash Hash::fromBytes(const std::array<uint8_t, SIZE>& bytes) {
    Hash hash;
    hash.bytes_ = bytes;
    return hash;
}

Hash Hash::fromHex(const std::string& hex) {
    std::string cleanHex = Hex::remove0xPrefix(hex);

    if (cleanHex.length() != SIZE * 2) {
        throw std::invalid_argument("Invalid hash length: expected 64 hex characters");
    }

    if (!Hex::isValidHex(cleanHex)) {
        throw std::invalid_argument("Invalid hex characters in hash");
    }

    auto bytes = Hex::toBytes(cleanHex);
    std::array<uint8_t, SIZE> arr;
    std::copy(bytes.begin(), bytes.end(), arr.begin());

    return fromBytes(arr);
}

std::array<uint8_t, Hash::SIZE> Hash::toBytes() const {
    return bytes_;
}

std::string Hash::toHex() const {
    return Hex::toHexString(bytes_);
}

bool Hash::operator==(const Hash& other) const {
    return bytes_ == other.bytes_;
}

bool Hash::operator!=(const Hash& other) const {
    return !(*this == other);
}

bool Hash::operator<(const Hash& other) const {
    return bytes_ < other.bytes_;
}

bool Hash::isZero() const {
    return std::all_of(bytes_.begin(), bytes_.end(), [](uint8_t b) { return b == 0; });
}

} // namespace quantaureum
